### PR TITLE
Added events to zenpackadapter

### PR DESCRIPTION
**Pull request by jeff-powell**
Add event processing to zenpackadapter as part of zminihub.
fixed a several issues causing some events and metrics to throw errors and be skipped
This addresses ZING-12577

### CODE REVIEW

**Comment by webhookbotuser** [comment]
Can one of the admins verify this patch?

**Comment by webhookbotuser** [comment]
Can one of the admins verify this patch?

**Comment by webhookbotuser** [comment]
Can one of the admins verify this patch?

File: Products/ZenHub/PBDaemon.py [hunk 1]
```diff
@@ -1210,6 +1217,12 @@ def buildOptions(self):
                                type='int',
                                help='Maximum number of events to queue')
 
+        self.parser.add_option('--maxqueueage',
```
Comment (joshw) [inline Products/ZenHub/PBDaemon.py:1220]: In general I try to keep changes to anything other than Products/ZenPackAdapter out of this branch.    I'd suggest getting this into develop (and merge it in here too), just to make it easier to keep things reconciled.

File: Products/ZenPackAdapter/utils/tales.py [hunk 2]
```diff
@@ -49,11 +49,8 @@ def talesEval(express, context, extra=None, skipfails=False):
         msg = "Error when processing tales expression %s on context %s : Exception Class %s Message: %s" % (express,
                                                                                                             context,
                                                                                                             type(e), e)
-        if skipfails:
-            res = express
-            log.debug(msg)
-        else:
-            raise InvalidTalesException(msg)
+        res = express
+        log.warn(msg)
```
Comment (joshw) [inline Products/ZenPackAdapter/utils/tales.py:53]: Is this still needed?  It looks like you're catching the InvalidTalesExpression and logging it elsewhere.

File: Products/ZenPackAdapter/zobject.py [hunk 3]
```diff
@@ -595,12 +595,14 @@ def device(self):
         return self
 
     def getDeviceComponents(self, collector=None):
+        components = []
         for component in self.getSubComponents():
             if collector is not None:
                 if collector in getattr(component, 'collectors', []):
-                    yield component
+                    components.append(component)
```
Comment (joshw) [inline Products/ZenPackAdapter/zobject.py:602]: not that I imagine it matters much, but why the change here?  this is slightly less efficient..

File: Products/ZenPackAdapter/zminievents.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
+##############################################################################
```
Comment (joshw) [inline Products/ZenPackAdapter/zminievents.py:1]: Is this used anywhere?

**Review by joshw — CHANGES_REQUESTED** [review]

File: Products/ZenHub/PBDaemon.py (hunk 1 above)
Comment (jeff-powell) [reply Products/ZenHub/PBDaemon.py:1220]: I can isolate it - I'm just not sure whether or not they'd want an age-limit added to the event queue. I can ask Jason about this too.

Without these changes, two-thirds of the buffer has to be filled before the daemon will send.  (The default buffer size is 5000.)  Event if we set a low (via configs,) a few events could wait around a long time before the daemon sends them.

File: Products/ZenPackAdapter/utils/tales.py (hunk 2 above)
Comment (jeff-powell) [reply Products/ZenPackAdapter/utils/tales.py:53]: I couldn't get zpa or the EMC zenpack to load correctly until i removed the "raise InvalideTalesException(msg)"; the exception was causing the zenpack to fail loading.  Hence, I updated this segment to operate similar to how res_manager operates; returning the expression unmodified.

From there, I never really re-visited this since ZPA has been running smooth for me since (so I could work on events.)

File: Products/ZenPackAdapter/zminievents.py (hunk 4 above)
Comment (jeff-powell) [reply Products/ZenPackAdapter/zminievents.py:1]: no, I should remove it.

File: Products/ZenPackAdapter/zobject.py (hunk 3 above)
Comment (jeff-powell) [reply Products/ZenPackAdapter/zobject.py:602]: when running zendmd on resmgr, a list with the components were provided
> dev1.getDeviceComponents()
[ compont_type_name_id_1,
compont_type_name_id_2,
etc,
] 

When running dmd on ZPA, the list is not generated, instead it's handle is returned
> dev1.getDeviceComponents()
(generator_object)

File: Products/ZenHub/PBDaemon.py (hunk 1 above)
Comment (jpeacock-zenoss) [reply Products/ZenHub/PBDaemon.py:1220]: This particular change looks reasonably isolated from the rest of the ZPA project to call it an enhancement for pushing events based on age (in addition to the queue size).

Comment (jeff-powell) [reply Products/ZenHub/PBDaemon.py:1220]: Interesting, that 5 second loop makes sense though.
I wonder if that particular code isn't running in ZPA, because when I working on events, they never were sent until I added this

Comment (jpeacock-zenoss) [reply Products/ZenHub/PBDaemon.py:1220]: The `pushEventsLoop` code should have been scheduled into the Twisted reactor when PBDaemon's `run` method is called.

Comment (jeff-powell) [reply Products/ZenHub/PBDaemon.py]: This is all correct now.  I updated monkey.py to correctly use and backed these changes to pbdaemon. 
`@monkeypatch(PBDaemon)
def pushEventsLoop(self):
    pass`

File: Products/ZenPackAdapter/zobject.py (hunk 3 above)
Comment (joshw) [reply Products/ZenPackAdapter/zobject.py:601]: Sure, that's typical of generators, but it's functionally equivalent (you can iterate on it, same as a list, or turn it into one  `list(dev1.getDeviceComponents())`for debugging)

**Review by joshw — APPROVED** [review]
